quipucords/tests/api/scantask/test_scantask.py

```diff
@@ -47,202 +48,195 @@ def scan_job(scan: Scan) -> ScanJob:
-        assert end_time.replace(microsecond=0) == task.end_time.replace(microsecond=0)
 
-    def test_scantask_increment(self, scan_job, source):
-        """Test scan task increment feature."""
-        task = ScanTask.objects.create(
-            job=scan_job,
-            source=source,
-            scan_type=ScanTask.SCAN_TYPE_CONNECT,
-            status=ScanTask.PENDING,
-        )
-        task.increment_stats(
-            "foo",
-            increment_sys_count=True,
-            increment_sys_scanned=True,
-            increment_sys_failed=True,
-            increment_sys_unreachable=True,
-        )
-        assert 1 == task.systems_count
-        assert 1 == task.systems_scanned
-        assert 1 == task.systems_failed
-        assert 1 == task.systems_unreachable
-        task.increment_stats(
-            "foo",
-            increment_sys_count=True,
-            increment_sys_scanned=True,
-            increment_sys_failed=True,
-            increment_sys_unreachable=True,
-        )
-        assert 2 == task.systems_count
-        assert 2 == task.systems_scanned
-        assert 2 == task.systems_failed
-        assert 2 == task.systems_unreachable
+@pytest.mark.django_db
+def test_scantask_increment_stats_multiple_in_parallel(scan_job, source):
+    """Test scan task increment two instances in parallel.
 
-    def test_scantask_increment_stats_multiple_in_parallel(self, scan_job, source):
-        """Test scan task increment two instances in parallel.
+    This is a special case test to help ensure that increment_stats is thread-safe.
+    We patch `refresh_from_db` with a mock object here because we don't want the
+    underlying functionality to rely on `refresh_from_db` completing immediately
+    before the update/save.
+    """
+    task_instance_a = ScanTask.objects.create(
+        job=scan_job,
+        source=source,
+        scan_type=ScanTask.SCAN_TYPE_CONNECT,
+        status=ScanTask.PENDING,
+    )
+    task_instance_b = ScanTask.objects.get(id=task_instance_a.id)
 
-        This is a special case test to help ensure that increment_stats is thread-safe.
-        We patch `refresh_from_db` with a mock object here because we don't want the
-        underlying functionality to rely on `refresh_from_db` completing immediately
-        before the update/save.
-        """
-        task_instance_a = ScanTask.objects.create(
-            job=scan_job,
-            source=source,
-            scan_type=ScanTask.SCAN_TYPE_CONNECT,
-            status=ScanTask.PENDING,
-        )
-        task_instance_b = ScanTask.objects.get(id=task_instance_a.id)
+    assert 0 == task_instance_a.systems_count
+    assert 0 == task_instance_b.systems_count
 
-        assert 0 == task_instance_a.systems_count
-        assert 0 == task_instance_b.systems_count
+    with (
+        patch.object(task_instance_a, "refresh_from_db"),
+        patch.object(task_instance_b, "refresh_from_db"),
+    ):
```

Comment:
I can see the original code had this patch, but do you know why? seems unnecessary 🤔

Reply:
No clue. I didn't dig into that, but let me see what happens if I yank them out…

Reply:
Oh. Well, there's this old docstring above it:

>    This is a special case test to help ensure that increment_stats is thread-safe.
>    We patch `refresh_from_db` with a mock object here because we don't want the
>    underlying functionality to rely on `refresh_from_db` completing immediately
>    before the update/save.

The test still passes when I commented out the patches and deindented the code it wrapped, but I'll just leave it like this. I don't feel like this is a _good_ test, but I'll dodge and say I'm trying to minimize scope creep. 🤷

Reply:
LOL I thought that writing style looked familiar. Guess who is on the blame for the original lines? 😅 https://github.com/quipucords/quipucords/commit/c21e53f6df4afe2bf3cde90f092abe0099d97c71

Reply:
lol

this caught my attention because, even though mocked, we still call the method anyways. Which is extremely counter intuitive to me. Looks like we are doing nothing when calling `refresh_from_db` later, so why do it?

Reply:
OH! I think I understand what I intended when I wrote that, and I think it still makes sense. `increment_stats` internally calls `refresh_from_db`. This test tries to simulate conditions of parallelism without actually running things in parallel. If you have two different instances of an object that both point to the same underlying database record, and you call `increment_stats` of each of them independently but around the same time, the underlying stats counter in the DB _should_ go from `0` to `2`. The code that existed _before_ I introduced a fix with that test was not thread-safe for cases like this, and the counter in the DB would go from `0` to only `1`, not the correct `2`.

The reason I mocked out `refresh_from_db` was to guarantee that the instances don't refresh their local copies of the underlying data until they've both completed their saves. If the second instance had refreshed its data before saving, then the simulated parallelism conditions would be invalid as the second might have "fixed" its in-memory counter to `1` when in a real parallel call it would still be `0`.

So, I believe the test should remain. I hope that clears it up a bit. 🙂